[Merged by Bors] - Support `system.in_schedule()` and `system.on_startup()`

JoJoJet opened the pull request:
# Objective

Support the following syntax for adding systems:

```rust
App::new()
    .add_system(setup.on_startup())
    .add_systems((
        show_menu.in_schedule(OnEnter(GameState::Paused)),
        menu_ssytem.in_set(OnUpdate(GameState::Paused)),
        hide_menu.in_schedule(OnExit(GameState::Paused)),
    ))
```

## Solution

Add the traits `IntoSystemAppConfig{s}`, which provide the extension methods necessary for configuring which schedule a system belongs to. These extension methods return `IntoSystemAppConfig{s}`, which `App::add_system{s}` uses to choose which schedule to add systems to.

---

## Changelog

+ Added the extension methods `in_schedule(label)` and  `on_startup()` for configuring the schedule a system belongs to.

## Future Work

* Replace all uses of `add_startup_system` in the engine.
* Deprecate this method

wilk10 commented:
Thanks a lot! The API looks great now, and much more digestible to first users (though it wasn't too difficult to grok before, just a bit more verbose). 

Too bad we lost `.on_update` along the way with https://github.com/bevyengine/bevy/pull/7667 . Is there a way to fit it back into this PR? That would give the most user-facing consistency imo.

JoJoJet commented:
> Too bad we lost .on_update along the way with #7667 . Is there a way to fit it back into this PR?

Yeah, we'll just have to decide whether `on_update` should live in `bevy_ecs` or `bevy_app` -- it *can* exist in the ECS, but it might feel out of place without the `on_enter` and `on_exit` variants.

Any thoughts?

alice-i-cecile commented:
My preference is definitely to have it live in `bevy_app`.

I'd prefer to add `on_update` back as part of this PR.

JoJoJet commented:
If the intention is to land this in 0.10, should I also remove newly-added methods like `add_system{s}_to_schedule`?

maniwani commented:
Having `on_update` would reintroduce #7634. It's not something we removed on a whim. We can't just bring it back. `OnUpdate` is a system set, not a schedule.

alice-i-cecile commented:
> should I also remove newly-added methods like add_system{s}_to_schedule

These are generally useful.

alice-i-cecile commented:
> 



> Having `on_update` would reintroduce #7634. It's not something we removed on a whim. We can't just bring it back. `OnUpdate` is a system set, not a schedule.

Fair 🤔 Okay, let's leave it out.

alice-i-cecile commented on crates/bevy_app/src/config.rs:
```diff
@@ -0,0 +1,341 @@
+    pub(crate) schedule: ScheduleMode,
+}
+
+pub(crate) enum ScheduleMode {
```
More docs here would be nice, this is internal but tricky.

alice-i-cecile reviewed (approved):
I have nits, but this is well made and a serious improvement to end user ergonomics.

Milestone is aspirational, but we should at least consider this as a last minute addition.

maniwani commented:
Can I get a word on the removal of those sealed traits? I vaguely understand, but just wanted to ask.

hymm commented:
Don't have time to review this right now, but I don't think we should reintroduce on_enter and on_exit without on_update. If we want to try to get this in we should just have in_schedule and push the other stuff into a different pr.

JoJoJet commented:
> Can I get a word on the removal of those sealed traits? I vaguely understand, but just wanted to ask.

They were never necessary in the first place. External implementations of these traits shouldn't result in anything weird or unsound. Plus the `SystemConfig{s}` structs can only be constructed within `bevy_ecs`, so the only way to return one is to forward to functions defined within `bevy_ecs` anyway.

JoJoJet commented:
> > should I also remove newly-added methods like add_system{s}_to_schedule
> 
> These are generally useful.

How so? `add_system_to_schedule(MyScheduleLabel, system)` seems strictly worse than `add_system(system.in_schedule(MyScheduleLabel))` IMO.

maniwani reviewed (approved):
This all looks OK to me. The lack of `on_update` is consistent with #7634, so I don't consider that a blocker.

(Personally, I'd like to move states out of `bevy_ecs` into their own plugin. They aren't an internal element of the scheduler.)

alice-i-cecile commented:
> > > should I also remove newly-added methods like add_system{s}_to_schedule
> > 
> > 
> > These are generally useful.
> 
> How so? `add_system_to_schedule(MyScheduleLabel, system)` seems strictly worse than `add_system(system.in_schedule(MyScheduleLabel))` IMO.

Ah, I'd missed that you'd added `.in_schedule` somehow. In that case yes, let's remove the `to_schedule` methods in this PR as well.

JoJoJet commented:
> Don't have time to review this right now, but I don't think we should reintroduce on_enter and on_exit without on_update. If we want to try to get this in for 0.10 we should just have in_schedule and push the other stuff into a different pr.

Let's try a quick poll. If you want `on_enter` and `on_exit` to stay, upvote this comment. If you want to remove them/save them for later, downvote this comment.

JoJoJet commented:
Alright, I've removed the state extension methods for now. We'll let the design for them cook a little more.

I also replaced every usage of `add_system{s}_to_schedule` with `.in_schedule` (thank the gods for regex).

I think we should also deprecate `add_startup_system{s}`, but I imagine that will be more controversial so I'll leave it for a follow-up.

alice-i-cecile commented:
> I think we should also deprecate add_startup_system{s}, but I imagine that will be more controversial so I'll leave it for a follow-up.

Agreed on both counts. With the state API cut, I've removed the controversial label and feel good about getting this in.

JoJoJet commented:
It seems to work fine on my machine. What does it mean when `CI/run-examples` passes but the bot says it fails?

JoJoJet commented:
I think the bot is just slow. `breakout` failed earlier, but it got resolved by the time the bot commented.

JoJoJet commented:
I realized that we probably don't want to expose `SystemConfigs as IntoIterator`, so I refactored the PR to not require that. I accidentally made the code a bit clearer as a result, too.

alice-i-cecile reviewed (approved):
bors r+
